src/Foundation/Commerce/code/Services/Billing/BillingService.cs

```diff
@@ -15,28 +15,255 @@
+                    this.AddPaymentOptions(result, cartResult.Cart);
+
+                    if (result.Success)
+                    {
+                        this.AddPaymentMethods(result, cartResult.Cart);
+
+                        if (result.Success)
+                        {
+                            this.AddPaymentClientToken(result);
+                        }
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        public Result<VoidResult> SetPaymentInfo(Address billingAddress, FederatedPaymentInfo federatedPayment)
+        {
+            Assert.ArgumentNotNull(billingAddress, nameof(billingAddress));
+            Assert.ArgumentNotNull(federatedPayment, nameof(federatedPayment));
+
+            var result = new Result<VoidResult>(new VoidResult());
+            var cartResult = this.cartManager.LoadCart(
+                this.storefrontContext.ShopName,
+                this.visitorContext.ContactId);
+
+            if (!cartResult.Success)
+            {
+                result.SetErrors(cartResult.SystemMessages.Select(m => m.Message).ToList());
+
+                return result;
+            }
+
+            cartResult = this.UpdateCartEmail(
+                cartResult.Cart,
+                string.IsNullOrWhiteSpace(billingAddress.Email)
+                    ? this.visitorContext.CurrentUser.Email
+                    : billingAddress.Email);
+
+            if (!cartResult.Success)
+            {
+                result.SetErrors(cartResult.SystemMessages.Select(m => m.Message).ToList());
+
+                return result;
+            }
+
+            if (cartResult.Cart.Payment != null && cartResult.Cart.Payment.Any())
+            {
+                cartResult = this.cartManager.RemovePaymentInfo(cartResult.Cart);
+
+                if (!cartResult.Success)
+                {
+                    result.SetErrors(cartResult.SystemMessages.Select(m => m.Message).ToList());
+
+                    return result;
+                }
+            }
+
+            var commerceParty = this.AddAddressToCart(cartResult.Cart, billingAddress);
+            var addPaymentInfoResult = this.AddPaymentInfoToCart(cartResult.Cart, federatedPayment, commerceParty);
+
+            if (!addPaymentInfoResult.Success)
+            {
+                result.SetErrors(addPaymentInfoResult.SystemMessages.Select(m => m.Message).ToList());
+
+                return result;
+            }
+
+            return result;
+        }
+
+        private CommerceParty AddAddressToCart(Cart cart, Address address)
+        {
+            var commerceParty = this.paymentMapper.Map<Address, CommerceParty>(address);
+
+            commerceParty.PartyId = Guid.NewGuid().ToString("N");
+            commerceParty.ExternalId = commerceParty.PartyId;
+
+            if (string.IsNullOrWhiteSpace(commerceParty.Name))
+            {
+                commerceParty.Name = $"billing{commerceParty.PartyId}";
```

Comment:
I think there is no restriction on commerce party name in Sitecore Commerce. I agree that it shouldn't be in Foundation.Commerce but I would like to discuss final approach first because such refactoring seems to be pretty huge for one task.